Aligning Wallet Instance Attestation with Attestation-Based Client Authentication via HTTP Headers (#327)
italia/eidas-it-wallet-docs: docs/en/wallet-attestation.rst

      "jwk": {
        "crv": "P-256",
        "kty": "EC",

>         "use": "sig",

Reviewer: no, don't add nothing else to this PR.

        "x": "4HNptI-xr2pjyRJKGMnz4WmdnQD_uJSq4R95Nj98b44",
        "y": "LIZnSB39vFJhYgS3k7jXE4r3-CoGFQwZtPBIRqpNlrg",

>         "kid": "vbeXJksM45xphtANnCiG6mCyuU4jfGNzopGuKvogg9c"

Reviewer: @fmarino-ipzs actually kid doesn't have sense in the cnf.jwk

The **token** endpoint (as defined in `RFC 7523 section 4`_) requires the following parameters
encoded in ``application/x-www-form-urlencoded`` format:

> * ``OAuth-Client-Attestation`` A JWT that conforms to the structure and syntax as defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;

Reviewer: ```suggestion
* ``OAuth-Client-Attestation`` A JWT that conforms to the structure and syntax defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
```

> * ``OAuth-Client-Attestation-PoP`` A JWT that conforms to the structure and syntax as defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;

Reviewer: ```suggestion
* ``OAuth-Client-Attestation-PoP`` A JWT that conforms to the structure and syntax defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
```

* ``grant_type`` set to ``urn:ietf:params:oauth:grant-type:jwt-bearer``;
* ``assertion`` containing the signed JWT of the Wallet Attestation Request.

> * ``OAuth-Client-Attestation and OAuth-Client-Attestation-PoP`` are both defined in this draft `specification <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_.

Reviewer: ```suggestion
* ``OAuth-Client-Attestation and OAuth-Client-Attestation-PoP`` are both defined in `OAuth Attestation Based Client Authentication <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_.
```

The **token** endpoint (as defined in `RFC 7523 section 4`_) requires the following parameters
encoded in ``application/x-www-form-urlencoded`` format:

* ``OAuth-Client-Attestation`` A JWT that conforms to the structure and syntax as defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
* ``OAuth-Client-Attestation-PoP`` A JWT that conforms to the structure and syntax as defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
* ``grant_type`` set to ``urn:ietf:params:oauth:grant-type:jwt-bearer``;
* ``assertion`` containing the signed JWT of the Wallet Attestation Request.

* ``OAuth-Client-Attestation and OAuth-Client-Attestation-PoP`` are both defined in this draft `specification <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_;

.. code-block:: http

    POST /token HTTP/1.1
    Host: wallet-provider.example.org
    Content-Type: application/x-www-form-urlencoded

>     OAuth-Client-Attestation: eyJhbGciOiAiRVMyNTYiLCJraWQiOiAidmJlWEprc004NXhwaHRBTm5DaUc2bUN5dVU0amZHTnpvcEd1S3ZvZ2c5YyIsInR5cCI6ICJ3YXIrand0In0.eyJpc3MiOiAiaHR0cHM6Ly93YWxsZXQtcHJvdmlkZXIuZXhhbXBsZS5vcmcvaW5zdGFuY2UvdmJlWEprc004NXhwaHRBTm5DaUc2bUN5dVU0amZHTnpvcEd1S3ZvZ2c5YyIsInN1YiI6ICJodHRwczovL3dhbGxldC1wcm92aWRlci5leGFtcGxlLm9yZy8iLCJjaGFsbGVuZ2UiOiAiNmVjNjkzMjQtNjBhOC00ZTViLWE2OTctYTc2NmQ4NTc5MGVhIiwiaGFyZHdhcmVfc2lnbmF0dXJlIjogIktvWklodmNOQVFjQ29JQXdnQUlCLnJlZGFjdGVkIiwiaW50ZWdyaXR5X2Fzc2VydGlvbiI6ICJvMk5tYlhSdlhCU2xZdEJwWVguLlJlZGFjdGVkIiwiaGFyZHdhcmVfa2V5X3RhZyI6ICJXUWh5RHltRktzUDk1aUZxcHpkRURXUzRsN2FWbmEyRm40SkNlV0hZdGJVPQ==",ImNuZiI6IHsiandrIjogeyJjcnYiOiAiUC0yNTYiLCJrdHkiOiAiRUMiLCJ1c2UiOiAic2lnIiwieCI6ICIxOE5wdEkteHIycGp5UkpLR01uejRXbWRuUURfdUpTcTRSOTVOajk4YjQ0IiwieSI6ICJMSVpud0IzOXZGSmpZZ1Mza3d4MUU0cjMtQ28HRlF3WnRQQklScW5ObHJnIiwia2lkIjogInZiZVhKa3NNTjV4cGh0QU5uQ2lHNm1DeXVVNGpmR05nWm9wR3VLdnVvZ2c5YyJ9fSwidmZfZm9ybWF0c19zdXBwb3J0ZWQiOiB7Imp3dF92Y19qc29uIjogeyJhbGdfdmFsdWVzX3N1cHBvcnRlZCI6IFsiRVMyNTZLIiwiRVMzODQiXX0sImp3dF92cF9qc29uIjogeyJhbGdfdmFsdWVzX3N1cHBvcnRlZCI6IFsi

Reviewer: remove `",`

  	OAuth-Client-Attestation-PoP: eyJhbGciOiJFUzI1NiJ9.eyJpc3MiOiJodHRwczovL2NsaWVudC5leGFtcGxlLmNvbSIsImF1ZCI6Imh0dHBzOi8vYXMuZXhhbXBsZS5jb20iLCJuYmYiOjEzMDA4MTU3ODAsImV4cCI6MTMwMDgxOTM4MH0.coB_mtdXwvi9RxSMzbIey8GVVQLv9qQrBUqmc1qj9Bs

    grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer
    &assertion=eyJhbGciOiJFUzI1NiIsImtpZCI6ImtoakZWTE9nRjNHeG...

**Steps 13-17**: The Wallet Provider's backend assesses the Wallet Attestation Request and issues a Wallet Attestation, if the requirements described below are satisfied:

>     1. It MUST check the Wallet Attestation Request contains all the defined header according to :ref:`Table of the Wallet Attestation Request Header <table_wallet_attestation_request_claim>`.

Reviewer: ```suggestion
    1. It MUST check the Wallet Attestation Request contains all the defined HTTP Request header parameters according to :ref:`Table of the Wallet Attestation Request Header <table_wallet_attestation_request_claim>`.
```

The **token** endpoint (as defined in `RFC 7523 section 4`_) requires the following parameters
encoded in ``application/x-www-form-urlencoded`` format:

* ``OAuth-Client-Attestation`` A JWT that conforms to the structure and syntax defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
* ``OAuth-Client-Attestation-PoP`` A JWT that conforms to the structure and syntax defined in the `OAuth 2.0 Attestation-Based Client Authentication` speficification;
* ``grant_type`` set to ``urn:ietf:params:oauth:grant-type:jwt-bearer``;
* ``assertion`` containing the signed JWT of the Wallet Attestation Request.

> * ``OAuth-Client-Attestation and OAuth-Client-Attestation-PoP`` are both defined in `OAuth Attestation Based Client Authentication <https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>`_.

Reviewer: don't use `<https://github.com/oauth-wg/draft-ietf-oauth-attestation-based-client-auth/blob/main/draft-ietf-oauth-attestation-based-client-auth.md>` but a ref to the defined specs

.. code-block:: http

    POST /token HTTP/1.1
    Host: wallet-provider.example.org
    Content-Type: application/x-www-form-urlencoded

>     OAuth-Client-Attestation: ICB7DQogICAgImFsZyI6ICJFUzI1NiIsDQogICAgImtpZCI6ICJ2YmVYSmtzTTQ1eHBodEFObkNpRzZtQ3l1VTRqZkdOem9wR3VLdm9nZzljIiwNCiAgICAidHlwIjogIndhcitqd3QiDQogIH0.ew0KICAiaXNzIjogImh0dHBzOi8vd2FsbGV0LXByb3ZpZGVyLmV4YW1wbGUub3JnL2luc3RhbmNlL3ZiZVhKa3NNNDV4cGh0QU5uQ2lHNm1DeXVVNGpmR056b3BHdUt2b2dnOWMiLA0KICAic3ViIjogImh0dHBzOi8vd2FsbGV0LXByb3ZpZGVyLmV4YW1wbGUub3JnLyIsDQogICJjaGFsbGVuZ2UiOiAiNmVjNjkzMjQtNjBhOC00ZTViLWE2OTctYTc2NmQ4NTc5MGVhIiwNCiAgImhhcmR3YXJlX3NpZ25hdHVyZSI6ICJLb1pJaHZjTkFRY0NvSUF3Z0FJQi4uLnJlZGFjdGVkIiwNCiAgImludGVncml0eV9hc3NlcnRpb24iOiAibzJObWJYUnZZWEJ3YkdVdFlYQndZWC4uLnJlZGFjdGVkIiwNCiAgImhhcmR3YXJlX2tleV90YWciOiAiV1FoeUR5bUZLc1A5NWlGcXB6ZEVEV1c0bDdhVm5hMkZuNEpDZVdIWXRiVT0iLA0KICAiY25mIjogew0KICAgICJqd2siOiB7DQogICAgICAiY3J2IjogIlAtMjU2IiwNCiAgICAgICJrdHkiOiAiRUMiLA0KICAgICAgIngiOiAiNEhOcHRJLXhyMnBqeVJKS0dNbno0V21kblFEX3VKU3E0Ujk1Tmo5OGI0NCIsDQogICAgICAieSI6ICJMSVpuU0IzOXZGSmhZZ1MzazdqWEU0cjMtQ29HRlF3WnRQQklScXBObHJnIiwNCiAgICAgICJraWQiOiAidmJlWEprc000NXhwaHRBTm5DaUc2bUN5dVU0amZHTnpvcEd1S3ZvZ2c5YyINCiAgICB9DQogIH0sDQogICJ2cF9mb3JtYXRzX3N1cHBvcnRlZCI6IHsNCiAgICAgICJqd3RfdmNfanNvbiI6IHsNCiAgICAgICAgImFsZ192YWx1ZXNfc3VwcG9ydGVkIjogWyJFUzI1NksiLCAiRVMzODQiXSwNCiAgICAgIH0sDQogICAgICAiand0X3ZwX2pzb24iOiB7DQogICAgICAgICJhbGdfdmFsdWVzX3N1cHBvcnRlZCI6IFsiRVMyNTZLIiwgIkVkRFNBIl0sDQogICAgICB9LA0KICAgIH0sDQogIH0sDQogICJpYXQiOiAxNjg2NjQ1MTE1LA0KICAiZXhwIjogMTY4NjY1MjMxNQ0KfQ

Reviewer: the payload contains `\r\n` and anso the kid in the CNF

this JWT should be recreated 

![immagine](https://github.com/user-attachments/assets/94be7995-3ffc-4517-b32d-9aba7227078e)

>     OAuth-Client-Attestation: ICB7DQogICAgImFsZyI6ICJFUzI1NiIsDQogICAgImtpZCI6ICJ2YmVYSmtzTTQ1eHBodEFObkNpRzZtQ3l1VTRqZkdOem9wR3VLdm9nZzljIiwNCiAgICAidHlwIjogIndhcitqd3QiDQogIH0.ew0KICAiaXNzIjogImh0dHBzOi8vd2FsbGV0LXByb3ZpZGVyLmV4YW1wbGUub3JnL2luc3RhbmNlL3ZiZVhKa3NNNDV4cGh0QU5uQ2lHNm1DeXVVNGpmR056b3BHdUt2b2dnOWMiLA0KICAic3ViIjogImh0dHBzOi8vd2FsbGV0LXByb3ZpZGVyLmV4YW1wbGUub3JnLyIsDQogICJjaGFsbGVuZ2UiOiAiNmVjNjkzMjQtNjBhOC00ZTViLWE2OTctYTc2NmQ4NTc5MGVhIiwNCiAgImhhcmR3YXJlX3NpZ25hdHVyZSI6ICJLb1pJaHZjTkFRY0NvSUF3Z0FJQi4uLnJlZGFjdGVkIiwNCiAgImludGVncml0eV9hc3NlcnRpb24iOiAibzJObWJYUnZZWEJ3YkdVdFlYQndZWC4uLnJlZGFjdGVkIiwNCiAgImhhcmR3YXJlX2tleV90YWciOiAiV1FoeUR5bUZLc1A5NWlGcXB6ZEVEV1c0bDdhVm5hMkZuNEpDZVdIWXRiVT0iLA0KICAiY25mIjogew0KICAgICJqd2siOiB7DQogICAgICAiY3J2IjogIlAtMjU2IiwNCiAgICAgICJrdHkiOiAiRUMiLA0KICAgICAgIngiOiAiNEhOcHRJLXhyMnBqeVJKS0dNbno0V21kblFEX3VKU3E0Ujk1Tmo5OGI0NCIsDQogICAgICAieSI6ICJMSVpuU0IzOXZGSmhZZ1MzazdqWEU0cjMtQ29HRlF3WnRQQklScXBObHJnIiwNCiAgICAgICJraWQiOiAidmJlWEprc000NXhwaHRBTm5DaUc2bUN5dVU0amZHTnpvcEd1S3ZvZ2c5YyINCiAgICB9DQogIH0sDQogICJ2cF9mb3JtYXRzX3N1cHBvcnRlZCI6IHsNCiAgICAgICJqd3RfdmNfanNvbiI6IHsNCiAgICAgICAgImFsZ192YWx1ZXNfc3VwcG9ydGVkIjogWyJFUzI1NksiLCAiRVMzODQiXSwNCiAgICAgIH0sDQogICAgICAiand0X3ZwX2pzb24iOiB7DQogICAgICAgICJhbGdfdmFsdWVzX3N1cHBvcnRlZCI6IFsiRVMyNTZLIiwgIkVkRFNBIl0sDQogICAgICB9LA0KICAgIH0sDQogIH0sDQogICJpYXQiOiAxNjg2NjQ1MTE1LA0KICAiZXhwIjogMTY4NjY1MjMxNQ0KfQ

Reviewer: ```suggestion
    OAuth-Client-Attestation: eyJ0eXAiOiJ3YWxsZXQtYXR0ZXN0YXRpb24rand0IiwiYWxnIjoiRVMyNTYiLCJraWQiOiJFVUcwRXZUVmlJNUVOWkF3VUNJVU1nUEFZOF9VSEluXzJISFpTMTd0X0MwIn0.eyJpc3MiOiAiaHR0cHM6Ly93YWxsZXQtcHJvdmlkZXIuZXhhbXBsZS5vcmcvaW5zdGFuY2UvdmJlWEprc000NXhwaHRBTm5DaUc2bUN5dVU0amZHTnpvcEd1S3ZvZ2c5YyIsICJzdWIiOiAiaHR0cHM6Ly93YWxsZXQtcHJvdmlkZXIuZXhhbXBsZS5vcmcvIiwgImNoYWxsZW5nZSI6ICI2ZWM2OTMyNC02MGE4LTRlNWItYTY5Ny1hNzY2ZDg1NzkwZWEiLCAiaGFyZHdhcmVfc2lnbmF0dXJlIjogIktvWklodmNOQVFjQ29JQXdnQUlCLi4ucmVkYWN0ZWQiLCAiaW50ZWdyaXR5X2Fzc2VydGlvbiI6ICJvMk5tYlhSdllYQndiR1V0WVhCd1lYLi4ucmVkYWN0ZWQiLCAiaGFyZHdhcmVfa2V5X3RhZyI6ICJXUWh5RHltRktzUDk1aUZxcHpkRURXVzRsN2FWbmEyRm40SkNlV0hZdGJVPSIsICJjbmYiOiB7Imp3ayI6IHsiY3J2IjogIlAtMjU2IiwgImt0eSI6ICJFQyIsICJ4IjogIjRITnB0SS14cjJwanlSSktHTW56NFdtZG5RRF91SlNxNFI5NU5qOThiNDQiLCAieSI6ICJMSVpuU0IzOXZGSmhZZ1MzazdqWEU0cjMtQ29HRlF3WnRQQklScXBObHJnIn19LCAidnBfZm9ybWF0c19zdXBwb3J0ZWQiOiB7Imp3dF92Y19qc29uIjogeyJhbGdfdmFsdWVzX3N1cHBvcnRlZCI6IFsiRVMyNTZLIiwgIkVTMzg0Il19LCAiand0X3ZwX2pzb24iOiB7ImFsZ192YWx1ZXNfc3VwcG9ydGVkIjogWyJFUzI1NksiLCAiRWREU0EiXX19LCAiaWF0IjogMTY4NjY0NTExNSwgImV4cCI6IDE2ODY2NTIzMTV9.ClxWia-9QHysYKW3WIEEFAFUIK1-seMHddhpW-2bKHRMqcLvRdck7Pvq0S-9uIaS5N-j7BPDGAJqQCR9A-haLw
```

      - **Description**
      - **Reference**
    * - **alg**
      - OPTIONAL. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST be one of the supported algorithms listed in the Section `Cryptographic Algorithms <algorithms.html>`_ and MUST NOT be set to ``none`` or any symmetric algorithm (MAC) identifier.
      - :rfc:`7516#section-4.1.1`.
    * - **kid**
      -  OPTIONAL. Unique identifier of the ``jwk`` used by the Wallet Provider to sign the Wallet Attestation, essential for matching the Wallet Provider's cryptographic public key needed for signature verification.
      - :rfc:`7638#section_3`.
    * - **typ**

>       -  OPTIONAL. It MUST be set to ``var+jwt``

Reviewer: it is not optional at all.

if it was MUST is therefore REQUIRED

      - **Description**
      - **Reference**
    * - **iss**

>       - MANDATORY. Identifier of the Wallet Provider concatenated with thumbprint of the JWK in the ``cnf`` parameter.

Reviewer: ```suggestion
      - REQUIRED. Identifier of the Wallet Provider concatenated with thumbprint of the JWK in the ``cnf`` parameter.
```

      - :rfc:`9126` and :rfc:`7519`.
    * - **aud**
      - OPTIONAL. It MUST be set to the identifier of the Wallet Provider.
      - :rfc:`9126` and :rfc:`7519`.
    * - **exp**

>       - MANDATORY. UNIX Timestamp with the expiry time of the JWT.

Reviewer: ```suggestion
      - REQUIRED. UNIX Timestamp with the expiry time of the JWT.
```

      - :rfc:`9126` and :rfc:`7519`.
    * - **iat**

>       - MANDATORY. UNIX Timestamp with the time of JWT issuance.

Reviewer: ```suggestion
      - REQUIRED. UNIX Timestamp with the time of JWT issuance.
```

      - :rfc:`9126` and :rfc:`7519`.
    * - **challenge**
      - OPTIONAL. Challenge data obtained from ``nonce`` endpoint
      -
    * - **hardware_signature**
      - OPTIONAL. The signature of ``client_data`` obtained using Cryptographic Hardware Key base64 encoded.
      -
    * - **integrity_assertion**
      - OPTIONAL. The integrity assertion obtained from the **Device Integrity Service** with the holder binding of ``client_data``.
      -
    * - **hardware_key_tag**
      - OPTIONAL. Unique identifier of the **Cryptographic Hardware Keys**
      -
    * - **cnf**

>       - MANDATORY. JSON object, containing the public part of an asymmetric key pair owned by the Wallet Instance.

Reviewer: ```suggestion
      - REQUIRED. JSON object, containing the public part of an asymmetric key pair owned by the Wallet Instance.
```
